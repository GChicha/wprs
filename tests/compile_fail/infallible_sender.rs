// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::mpsc;
use std::sync::mpsc::Receiver;
use std::sync::mpsc::Sender;

use wprs::channel_utils::InfallibleSender;

struct ChannelParent(Sender<()>, Receiver<()>);
impl ChannelParent {
    fn new() -> Self {
        let (tx, rx) = mpsc::channel();
        Self(tx, rx)
    }

    fn sender<'a>(&'a self) -> Sender<()> {
        self.0.clone()
    }

    fn infallible_sender(&self) -> InfallibleSender<Sender<()>> {
        InfallibleSender::new(self.0.clone(), self)
    }
}

fn main() {
    {
        let _sender = {
            let channel_parent = ChannelParent::new();
            channel_parent.sender()
        };
    }

    {
        let channel_parent = ChannelParent::new();
        let _infallible_sender = channel_parent.infallible_sender();
    }

    {
        let _infallible_sender = {
            let channel_parent = ChannelParent::new();
            channel_parent.infallible_sender()
        };
    }
}
