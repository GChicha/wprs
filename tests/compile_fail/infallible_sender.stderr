error[E0597]: `channel_parent` does not live long enough
  --> tests/compile_fail/infallible_sender.rs:53:13
   |
51 |         let _infallible_sender = {
   |             ------------------ borrow later stored here
52 |             let channel_parent = ChannelParent::new();
   |                 -------------- binding `channel_parent` declared here
53 |             channel_parent.infallible_sender()
   |             ^^^^^^^^^^^^^^ borrowed value does not live long enough
54 |         };
   |         - `channel_parent` dropped here while still borrowed
